### PR TITLE
[python] Create model level virtualenv

### PR DESCRIPTION
**Pull request by xyang16**
## Description ##

Brief description of what this PR is about

Create model-specific virtualenv. I discussed with Frank about two options:

1. Option 1: Create a virtual environment on the fly for each model.
2. Option 2: Pre-create a pool of n virtual environments and allocate one to each model.

In my experiment, the time creating the venv on the fly is not noticeable. So I had this PR for option 1.

### CODE REVIEW

File: engines/python/src/main/java/ai/djl/python/engine/PyEnv.java [hunk 1]
```diff
@@ -157,6 +157,51 @@ public Map<String, String> getInitParameters() {
         return initParameters;
     }
 
+    /**
+     * Create python virtual environment if needed.
+     *
+     * @param name the virtual environment name
+     */
+    public synchronized void createVirtualEnv(String name) {
+        Path path = Utils.getVenvDir().resolve(name).toAbsolutePath();
```
Comment (frankfliu) [inline engines/python/src/main/java/ai/djl/python/engine/PyEnv.java:166]: Utils.getVenvDir() doesn't seems belong to DJL, we should keep it in Python engine

File: engines/python/src/main/java/ai/djl/python/engine/PyProcess.java [hunk 2]
```diff
@@ -124,6 +124,7 @@ synchronized void startPythonProcess() {
             int id = restartCount.get();
             int port = connections.get(0).getPort();
             logger.info("Start process: {} - retry: {}", port, id);
+            pyEnv.createVirtualEnv(model.getName());
```
Comment (frankfliu) [inline engines/python/src/main/java/ai/djl/python/engine/PyProcess.java:127]: Using venv should be disabled by default.

File: engines/python/src/main/java/ai/djl/python/engine/PyEnv.java (hunk 1 above)
Comment (xyang16) [reply engines/python/src/main/java/ai/djl/python/engine/PyEnv.java:166]: I forgot to create PR to DJL. I added it to api

File: serving/docker/scripts/create_venv.sh [hunk 3]
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/env bash
+set -ex
+VENV_DIR=$1
+VENV_NAME=$2
+
+python3 -m venv ${VENV_DIR}/${VENV_NAME} --system-site-packages
+if [[ ${VENV_NAME} == "tgi" ]]; then
+    git clone https://github.com/OlivierDehaene/transformers.git
```
Comment (maaquib) [inline serving/docker/scripts/create_venv.sh:8]: This is no longer required https://github.com/huggingface/text-generation-inference/blob/abd58ff82c37d5e4f131abdac3d298927a815604/server/pyproject.toml#L29
Instead we should install [flash-attention](https://github.com/huggingface/text-generation-inference/blob/main/server/Makefile-flash-att)

File: engines/python/src/main/java/ai/djl/python/engine/PyEnv.java (hunk 1 above)
Comment (xyang16) [reply engines/python/src/main/java/ai/djl/python/engine/PyEnv.java]: Changed.

File: engines/python/src/main/java/ai/djl/python/engine/PyProcess.java (hunk 2 above)
Comment (xyang16) [reply engines/python/src/main/java/ai/djl/python/engine/PyProcess.java]: Done.

File: engines/python/src/main/java/ai/djl/python/engine/PyEnv.java [hunk 4]
```diff
@@ -163,6 +163,60 @@ public Map<String, String> getInitParameters() {
...
+     *
+     * @param name the virtual environment name
+     */
+    public synchronized void createVirtualEnv(String name) {
+        if (venvCreated) {
+            return;
+        }
+        Path path = getVenvDir().resolve(name).toAbsolutePath();
+        if (path.toFile().exists()) {
+            logger.debug("Virtual environment already exists at {}.", path);
```
Comment (frankfliu) [inline engines/python/src/main/java/ai/djl/python/engine/PyEnv.java:177]:
```suggestion
            logger.info("Virtual environment already exists at {}.", path);
```

File: engines/python/src/main/java/ai/djl/python/engine/PyEnv.java [hunk 5]
```diff
@@ -163,6 +163,60 @@ public Map<String, String> getInitParameters() {
...
+     * Create python virtual environment if needed.
+     *
+     * @param name the virtual environment name
+     */
+    public synchronized void createVirtualEnv(String name) {
+        if (venvCreated) {
+            return;
+        }
+        Path path = getVenvDir().resolve(name).toAbsolutePath();
+        if (path.toFile().exists()) {
```
Comment (frankfliu) [inline engines/python/src/main/java/ai/djl/python/engine/PyEnv.java:176]:
```suggestion
        if (Files.exists(path)) {
```

File: engines/python/src/main/java/ai/djl/python/engine/PyEnv.java [hunk 6]
```diff
@@ -163,6 +163,60 @@ public Map<String, String> getInitParameters() {
...
+                logger.warn("Failed to create virtual environment with error code: {}", ret);
+                logger.warn("{}", logOutput);
+            }
+        } catch (IOException | InterruptedException e) {
+            logger.warn("Python virtual failed.", e);
+        }
+    }
+
+    /**
+     * Delete python virtual environment.
```
Comment (frankfliu) [inline engines/python/src/main/java/ai/djl/python/engine/PyEnv.java:206]:
```suggestion
     * Deletes python virtual environment.
```

File: engines/python/src/main/java/ai/djl/python/engine/PyEnv.java [hunk 7]
```diff
@@ -163,6 +163,60 @@ public Map<String, String> getInitParameters() {
         return initParameters;
     }
 
+    /**
+     * Create python virtual environment if needed.
```
Comment (frankfliu) [inline engines/python/src/main/java/ai/djl/python/engine/PyEnv.java:167]:
```suggestion
     * Creates python virtual environment if needed.
```

File: engines/python/src/main/java/ai/djl/python/engine/PyEnv.java [hunk 8]
```diff
@@ -163,6 +163,60 @@ public Map<String, String> getInitParameters() {
...
+     * Delete python virtual environment.
+     *
+     * @param name the virtual environment name
+     */
+    public synchronized void deleteVirtualEnv(String name) {
+        if (!venvCreated) {
+            return;
+        }
+        Path path = getVenvDir().resolve(name).toAbsolutePath();
+        if (path.toFile().exists()) {
```
Comment (frankfliu) [inline engines/python/src/main/java/ai/djl/python/engine/PyEnv.java:215]: no need to check

File: engines/python/src/main/java/ai/djl/python/engine/PyEnv.java [hunk 9]
```diff
@@ -163,6 +163,60 @@ public Map<String, String> getInitParameters() {
...
+    /**
+     * Delete python virtual environment.
+     *
+     * @param name the virtual environment name
+     */
+    public synchronized void deleteVirtualEnv(String name) {
+        if (!venvCreated) {
+            return;
+        }
+        Path path = getVenvDir().resolve(name).toAbsolutePath();
```
Comment (frankfliu) [inline engines/python/src/main/java/ai/djl/python/engine/PyEnv.java:214]:
```suggestion
        Path path = getVenvDir().resolve(name);
```

File: engines/python/src/main/java/ai/djl/python/engine/PyProcess.java [hunk 10]
```diff
@@ -122,6 +122,9 @@ synchronized void startPythonProcess() {
             int id = restartCount.get();
             int port = connections.get(0).getPort();
             logger.info("Start process: {} - retry: {}", port, id);
+            if (pyEnv.isEnableVenv()) {
```
Comment (frankfliu) [inline engines/python/src/main/java/ai/djl/python/engine/PyProcess.java:125]: create/delete venv should happen at model loading and close time, not at StartPyProcess time

one model may have multiple workers.

File: engines/python/src/main/java/ai/djl/python/engine/PyEnv.java [hunk 11]
```diff
@@ -57,17 +58,16 @@ public class PyEnv {
...
+    private boolean venvCreated;
+
     /**
      * Constructs a new {@code PyEnv} instance.
      *
      * @param mpiMode true to use MPI launcher
      */
     public PyEnv(boolean mpiMode) {
         this.mpiMode = mpiMode;
-        pythonExecutable = Utils.getenv("PYTHON_EXECUTABLE");
```
Comment (frankfliu) [inline engines/python/src/main/java/ai/djl/python/engine/PyEnv.java:67]: With your implementation, no need to lazy get

File: engines/python/src/main/java/ai/djl/python/engine/PyEnv.java [hunk 12]
```diff
@@ -163,6 +163,59 @@ public Map<String, String> getInitParameters() {
...
+     * @param name the virtual environment name
+     */
+    public synchronized void createVirtualEnv(String name) {
+        if (venvCreated) {
+            return;
+        }
+        Path path = getVenvDir().resolve(name).toAbsolutePath();
+        if (Files.exists(path)) {
+            logger.info("Virtual environment already exists at {}.", path);
+            venvCreated = true;
```
Comment (frankfliu) [inline engines/python/src/main/java/ai/djl/python/engine/PyEnv.java:178]:
```suggestion
            setPythonExecutable(path.resolve("bin").resolve("python").toString());
            venvCreated = true;
```

File: engines/python/src/main/java/ai/djl/python/engine/PyEnv.java [hunk 13]
```diff
@@ -163,6 +163,59 @@ public Map<String, String> getInitParameters() {
...
+                logger.info("Python virtual environment created successfully at {}!", path);
+                logger.debug("{}", logOutput);
+                setPythonExecutable(path.resolve("bin").resolve("python").toString());
+                venvCreated = true;
+            } else {
+                logger.warn("Failed to create virtual environment with error code: {}", ret);
+                logger.warn("{}", logOutput);
+            }
+        } catch (IOException | InterruptedException e) {
+            logger.warn("Python virtual failed.", e);
```
Comment (frankfliu) [inline engines/python/src/main/java/ai/djl/python/engine/PyEnv.java:202]: We should fail the model loading

File: engines/python/src/main/java/ai/djl/python/engine/PyEnv.java [hunk 14]
```diff
@@ -163,6 +163,59 @@ public Map<String, String> getInitParameters() {
...
+                logOutput = Utils.toString(is);
+            }
+            int ret = process.waitFor();
+            if (ret == 0) {
+                logger.info("Python virtual environment created successfully at {}!", path);
+                logger.debug("{}", logOutput);
+                setPythonExecutable(path.resolve("bin").resolve("python").toString());
+                venvCreated = true;
+            } else {
+                logger.warn("Failed to create virtual environment with error code: {}", ret);
```
Comment (frankfliu) [inline engines/python/src/main/java/ai/djl/python/engine/PyEnv.java:198]: We should fail here

Comment (frankfliu) [reply engines/python/src/main/java/ai/djl/python/engine/PyEnv.java:198]: We need add a unit test

File: engines/python/src/main/java/ai/djl/python/engine/PyProcess.java (hunk 10 above)
Comment (xyang16) [reply engines/python/src/main/java/ai/djl/python/engine/PyProcess.java]: Changed.

File: engines/python/src/main/java/ai/djl/python/engine/PyEnv.java (hunk 14 above)
Comment (xyang16) [reply engines/python/src/main/java/ai/djl/python/engine/PyEnv.java]: throwed RuntimeException

Comment (xyang16) [reply engines/python/src/main/java/ai/djl/python/engine/PyEnv.java]: Added unit test

File: engines/python/src/main/java/ai/djl/python/engine/PyEnv.java (hunk 11 above)
Comment (xyang16) [reply engines/python/src/main/java/ai/djl/python/engine/PyEnv.java]: Changed.

**Review by frankfliu — APPROVED** [review]

**Review by lanking520 — APPROVED** [review]

**Pull request merged**